### PR TITLE
Document BLOOM lm_logits original training behavior

### PR DESCRIPTION
**Pull request by shijie-wu**
# What does this PR do?

~Compute softmax of Bloom in fp32 during half percision~

Document BLOOM lm_logits original training behavior

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@thomasw21 @stas00 @TevenLeScao

### CODE REVIEW

**Comment by sgugger** [comment]
cc @thomasw21 and @younesbelkada but I don't think this is necessary as bfloat16 is more numerically stable than float16?

**Comment by shijie-wu** [comment]
we noticed considerable performance difference between softmax in bf16 and softmax in fp32 in an internal bloom-based model. since during training the softmax is conducted in fp32, this change better reflects the model behavior in megatron-deepspeed.

**Comment by philschmid** [comment]
@shijie-wu can you please share a bit more about how you run the model? @thomasw21 shared with me that you are using SageMaker Model Parallelism (SMP)

**Comment by shijie-wu** [comment]
We redid the experiment regarding fp32 in SA. It seems like the gap we observed earlier is caused by some other issues. I have remove that part of the PR. After discussion with @thomasw21, instead of enforcing a type conversion, this PR will instead document the original behavior so that advanced users could recover it.

File: src/transformers/models/bloom/modeling_bloom.py [hunk 1]
```diff
@@ -871,6 +873,9 @@ def forward(
         hidden_states = transformer_outputs[0]
 
         lm_logits = self.lm_head(hidden_states)
+        # the original softmax is computed in fp32 during training
+        # https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/09a35f53abac96903fee50787426b7ee5f63fc62/megatron/model/gpt_model.py#L289-L29
+        # to replicate the same behavior as training, make sure to cast lm_logits to fp32
```
Comment (sgugger) [inline src/transformers/models/bloom/modeling_bloom.py:878]: You could add the line of code to uncomment as well here.

**Review by sgugger — COMMENTED** [review]
Thanks for iterating! I left some comments.

File: src/transformers/models/bloom/modeling_bloom.py (hunk 1 above)
Comment (shijie-wu) [reply src/transformers/models/bloom/modeling_bloom.py:878]: do you think we still need this comment section if we put it in the tips section of model card?

Comment (sgugger) [reply src/transformers/models/bloom/modeling_bloom.py:878]: Not necessarily!
